enhancements/single-node-production-deployment-approach.md

```diff
@@ -0,0 +1,484 @@
+introduced](https://github.com/openshift/enhancements/pull/555), all
+teams developing OpenShift components will need to consider how their
+components should be configured when deployed and used when the
+high-availability capability is disabled.
+
+In the telco RAN use case, high-availability is typically achieved by
+having multiple sites provide coverage to overlapping geographic
+areas. Therefore, use of cluster-based high-availability features will
+be limited (for example, by running a single API service).
+
+Failures in edge deployments are frequently resolved by re-imaging or
+physically replacing the entire host. Combining this fact with the
+previous observation about the approach to providing highly-available
+services lets us draw the conclusion that in-place upgrades do not
+need to be supported by the first iteration of this configuration.
+
+In a multi-node cluster, operators may safely trigger rolling reboots
+or restarts of the control plane services without taking the cluster
+offline. In a single-node deployment, any such restart would take some
+or all of the services offline. The deployment or upgrade of a cluster
+may include several host reboots, clustered together in time, until
+the operators stabilize. Users should wait for this period of
+instability before installing their workloads.
+
+The single-node configuration of OpenShift will be sufficiently
+generic to cater to a variety of edge computing use cases. As such,
+OpenShift's usual cluster configuration mechanisms will be favored
+where there is a likelihood that there will be edge computing use case
+with differing requirements.  For example, there is no expectation
+that single-node deployments will use the real-time kernel by
+default -- this will continue to be a `MachineConfig` choice as per
+[enhancements/support-for-realtime-kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md). *See open questions*
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+deployment, with some limitations (see non-goals above). The user
+will not require special support exceptions to receive technical assistance
+for the features supported by the configuration.
+
+### Implementation Details/Notes/Constraints
+
+Some OpenShift components require a minimum of 2 or 3 nodes. The new
+cluster capabilities API will be used to configure these components as
+appropriate for a single node.
+
+#### installer
+
+We will add logic to the installer to set the new
+`highAvailabilityMode` status field of the `infrastructure` API
+resource. The will value be set to `Full` by default and `None` when
+the number of replicas for the control plane is `1`. See the [cluster
+high-availability mode
+API](https://github.com/openshift/enhancements/pull/555) enhancement
+for details.
+
+#### machine-config-operator
+
```

Comment:
Can we add something explicit like:

The mco will support feature parity with OpenShift on a single node, excluding high availability and upgrades. Non-user initiated tasks will be non-disruptive, but user initiated tasks will be disruptive.

I think this should also be added to the graduation criteria below.

cc: @sinnykumari @yuqi-zhang @darkmuggle

Reply:
> Can we add something explicit like:
> 
> The mco will support feature parity with OpenShift on a single node, excluding high availability and upgrades. 

This is really implicit in saying OpenShift as a whole will do those things, so I'm reluctant to call the MCO out as a special case. Do you feel like there's something different in the expectations for MCO compared to the other components, though?

> Non-user initiated tasks will be non-disruptive, but user initiated tasks will be disruptive.

I know the MCO team is working on this, and it will benefit users of single-node deployments, but I don't think we need to consider it a blocker for the dev-preview version of single-node deployments described here. I can include it, but it actually raises the expectations of what would be delivered beyond what has been asked. Do we want to do that?

Reply:
@dhellmann the MCO is wanting to set the guard rails and explicitly set the expectation for the users. And while its implicit, the MCO team wants to communicate the implicit explicitly.

> I know the MCO team is working on this, and it will benefit users of single-node deployments, but I don't think we need to consider it a blocker for the dev-preview version of single-node deployments described here. 

As I understand this, it's a requirement on the MCO for Dev preview. However, I think it should be a graduation criterion for moving from Dev to Tech preview.

Reply:
I agree on this the first point. The wording that @kikisdeliveryservice mentioned was developed by the MCO team to fully define the scope of their work, but I agree that that's more appropriate to be included in our project tracking efforts than in this technical overview.

On the second point, I think the issue we are trying to highlight is that MCO _will_ make disruptive changes to the cluster. There is no plan in place to add any safeguards or otherwise change the behavior of MCO to ensure that reboots don't happen. There is a justified fear that, once delivered, the MCO team is going to bear the brunt of customer complaints about how workloads were disrupted shortly after configuration changes had been requested. We just want to be very clear that changes to Machine Configs, whether made by an Operator or admin, will trigger a reboot (aside from a few, select cases).

Reply:
> I agree on this the first point. The wording that @kikisdeliveryservice mentioned was developed by the MCO team to fully define the scope of their work, but I agree that that's more appropriate to be included in our project tracking efforts than in this technical overview.

Yeah, repeating it here is like saying "the MCO team will implement this enhancement." We can do that, it just feels a little redundant, so I was looking for either something more specific or some stronger reasoning about why it needs to be repeated here after having been stated above.

> On the second point, I think the issue we are trying to highlight is that MCO _will_ make disruptive changes to the cluster. There is no plan in place to add any safeguards or otherwise change the behavior of MCO to ensure that reboots don't happen. There is a justified fear that, once delivered, the MCO team is going to bear the brunt of customer complaints about how workloads were disrupted shortly after configuration changes had been requested. We just want to be very clear that changes to Machine Configs, whether made by an Operator or admin, will trigger a reboot (aside from a few, select cases).

I agree that the way we triage reboot-related failures is far from optimal and especially bad for the MCO team.

How about if we add some wording to the risks section to cover the concerns about disruptive changes? Something like "Single-node deployments of OpenShift necessarily will be prone to disruption when `MachineConfig` changes are applied."? Or maybe taking some of what's already there in the "Control plane rollouts" section and repeating it for `MachineConfig`? And then when the time comes we can also work with the docs team to make sure the warning is in the release notes and any other relevant user-facing documentation.

Reply:
I like all three of those suggestions.

Reply:
I'll open another PR to add this info and tag all of you for reviews.

Reply:
https://github.com/openshift/enhancements/pull/587